subprojects/enterprise/src/main/java/org/gradle/internal/enterprise/impl/GradleEnterpriseAutoAppliedPluginRegistry.java

```diff
@@ -0,0 +1,82 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.enterprise.impl;
+
+import org.gradle.StartParameter;
+import org.gradle.api.Project;
+import org.gradle.api.artifacts.ModuleIdentifier;
+import org.gradle.api.artifacts.ModuleVersionSelector;
+import org.gradle.api.initialization.Settings;
+import org.gradle.api.internal.StartParameterInternal;
+import org.gradle.api.internal.artifacts.DefaultModuleIdentifier;
+import org.gradle.api.internal.artifacts.DefaultModuleVersionSelector;
+import org.gradle.api.invocation.Gradle;
+import org.gradle.plugin.management.internal.DefaultPluginRequest;
+import org.gradle.plugin.management.internal.PluginRequestInternal;
+import org.gradle.plugin.management.internal.PluginRequests;
+import org.gradle.plugin.management.internal.autoapply.AutoAppliedGradleEnterprisePlugin;
+import org.gradle.plugin.management.internal.autoapply.AutoAppliedPluginRegistry;
+
+import static org.gradle.initialization.StartParameterBuildOptions.BuildScanOption;
+
+public class GradleEnterpriseAutoAppliedPluginRegistry implements AutoAppliedPluginRegistry {
+
+    private final GradleEnterprisePluginAutoAppliedStatus autoAppliedStatus;
+
+    public GradleEnterpriseAutoAppliedPluginRegistry(GradleEnterprisePluginAutoAppliedStatus autoAppliedStatus) {
+        this.autoAppliedStatus = autoAppliedStatus;
+    }
+
+    @Override
+    public PluginRequests getAutoAppliedPlugins(Project target) {
+        return PluginRequests.EMPTY;
+    }
+
+    @Override
+    public PluginRequests getAutoAppliedPlugins(Settings target) {
+        if (((StartParameterInternal) target.getStartParameter()).isUseEmptySettings() || !shouldApplyGradleEnterprisePlugin(target)) {
+            return PluginRequests.EMPTY;
+        } else {
+            return PluginRequests.of(createGradleEnterprisePluginRequest());
+        }
+    }
+
+    private boolean shouldApplyGradleEnterprisePlugin(Settings settings) {
+        Gradle gradle = settings.getGradle();
+        StartParameter startParameter = gradle.getStartParameter();
+        return startParameter.isBuildScan() && gradle.getParent() == null;
+    }
+
+    private PluginRequestInternal createGradleEnterprisePluginRequest() {
+        ModuleIdentifier moduleIdentifier = DefaultModuleIdentifier.newId(AutoAppliedGradleEnterprisePlugin.GROUP, AutoAppliedGradleEnterprisePlugin.NAME);
+        ModuleVersionSelector artifact = DefaultModuleVersionSelector.newSelector(moduleIdentifier, AutoAppliedGradleEnterprisePlugin.VERSION);
+        return new DefaultPluginRequest(
+            AutoAppliedGradleEnterprisePlugin.ID,
+            AutoAppliedGradleEnterprisePlugin.VERSION,
+            true,
+            null,
+            getScriptDisplayName(),
+            artifact,
+            null,
+            autoAppliedStatus::markAsAutoApplied
+        );
```

Comment:
I don’t have much context here, but what’s strange to me is that we need to use this kind of listener-style approach and then mutate some internal state (`markAsAutoApplied::markAsAutoApplied()`).

IIUC, `getAutoAppliedPlugins()` returns a GE plugin request to be auto-applied, but we should only report the plugin as auto-applied _if_ the application was actually successful, right?

If that's the case, can't we instead look for the list of successfully applied plugins, and see if the GE plugin is among them?

Reply:
No, because it being successfully applied does not mean it was auto-applied. We need to differentiate whether it was applied only because of `--scan` and not declared as a plugin in the settings script.

Reply:
There's two pieces of information: whether the plugin was among the auto-applied and whether the plugin was successfully applied. We could ask for both and only mark it as auto-applied if both checks pass.

Reply:
We could but it's irrelevant here since this information is only for consumption in the GE plugin which won't be called unless it's successfully applied.

Reply:
My point is that if we can keep `PluginRequestInternal` simpler, then we should consider doing so. This type is not GE-only, right?

Reply:
`PluginRequestInternal` would still have to store whether its origin was auto-application so I don't think that would be simpler.

Reply:
Replaced in the latest set of commits.